### ISSUE TITLE
Replace PhantomJS with Puppeteer

### ISSUE THREAD

**Issue by fogrew** [svg-sprite/svg-sprite#254]
[I see](https://github.com/jkphl/gulp-svg-sprite/issues/69) [that](https://github.com/jkphl/svg-sprite/issues/100) [many](https://github.com/jkphl/svg-sprite/issues/239) [people](https://github.com/jkphl/svg-sprite/issues/211) [have](https://github.com/jkphl/svg-sprite/issues/222) [problems](https://github.com/jkphl/svg-sprite/pull/244) with the complexity of installing PhantomJS.

Also, the project is officially dead. Do you think it's a good idea to replace PhantomJS with [Puppeteer](https://github.com/GoogleChrome/puppeteer)?
Find out the real dimensions of SVG without width and height there is not difficult.
Puppeteer is entirely node.js-compatible, might correctly installed almost everywhere. It indeed is more compatible than PhantomJS. The browser in it can be run with parameters for greater compatibility (disabling hardware acceleration, for example).

Do you think it's a good idea to replace PhantomJS with [Puppeteer](https://github.com/GoogleChrome/puppeteer)?

**Comment by zshnr** [comment]
I agree. Happy to help with a PR to do the replacement :)

**Comment by gamebox** [comment]
I agree with this as well, and would also be willing to put in a PR.

**Comment by codinronan** [comment]
Go for it, we are currently moving 100% to puppeteer and have already set up the build stack for it, so this would be a great move.

**Comment by DanielRuf** [comment]
PhantomJS is officially not developed or maintained anymore. For what is PhantomJS used / needed here?

**Comment by jkphl** [comment]
@DanielRuf See #100

**Comment by DanielRuf** [comment]
https://www.npmjs.com/package/sharp gets some good results.

```
const sharp = require('sharp');

const image = sharp('test/fixture/svg/special/file.svg');
image
  .metadata()
  .then(function(metadata) {
    console.log(metadata)
  })
```

**Comment by jkphl** [comment]
@DanielRuf Sounds promising, thanks! I'll look into this as soon as i find some time. Getting rid of PhantomJS is long overdue ...

**Comment by tkolsen** [comment]
Just wondering if anything is happening with this issue?
I see that phantomjs-prebuilt is stil a dependency.

**Comment by vyteniskuciauskas-home24** [comment]
Also move this dependency to the devDependencies. If I install svg-sprite to my project I will download phantomjs-prebuild and this is not required to work. Also download takes time.

**Comment by steffex** [comment]
@jkphl is there a way to disable this function with configuration? In my case I have only SVG's with the dimension properties set. It would be acceptable to be able to disable the phantomjs part, so it won't be loaded when using svg-sprite.

**Comment by DanielRuf** [comment]
You can add some `phantomjs` binary on your path.

See https://www.npmjs.com/package/phantomjs-prebuilt#continuous-integration

Or for example create an alias for `phantomjs` in your env which outpufs some string / version (mock).

**Comment by DanielRuf** [comment]
Also see https://github.com/DanielRuf/phantomjs-mocked which works great fir phantomjs 2.1.1

**Comment by DanielRuf** [comment]
Also see https://github.com/DanielRuf/phantomjs-mocked which works great for phantomjs 2.1.1

**Comment by mojoaxel** [comment]
I've created a pull-request (#284) that removes `phantomjs` in favor of `pupeteer`. It would be nice If you folks could test the changes in your setup. Just replace the `svg-sprite` dependency with the custom branch in your package.json:

```json
  "svg-sprite": "mojoaxel/svg-sprite#puppeteer",
```

**Comment by tsal** [comment]
We now have a 2-year old PR to remove phantomjs - a project abandoned and recommended to not-be-used by the original developers. PLEASE can we get this merged?

**Comment by joshdcomp** [comment]
It's worth noting this is now starting to cause issues for folks upgrading to Node 12—is this repo/package still maintained?

**Comment by rainfore** [comment]
I have a forked version, just removed PhantomJS.
@vusion/svg-sprite@1.6.0, https://github.com/vusion/svg-sprite

A little difference, not computing in headless browser if no width and height

**Comment by DanielRuf** [comment]
Eh, this breaks things.

https://github.com/vusion/svg-sprite/commit/27358ef1c3de28c61b3c96bc6c5f167dcdcd0899

Is it replaced or removed? Why not pick up this PR and use this as base?

**Comment by rainfore** [comment]
It takes us a lot of time to install a headless browser for a small loader when updating npm packages in our projects.
PhantomJS just plays a little role in this loader.

**Comment by DanielRuf** [comment]
Well, why not puppeteer and install it as peerDependency, create a postinstall script (combined with env variables) or mock it?

https://github.com/DanielRuf/phantomjs-mocked

Removing is not a solition ;-)

**Comment by DanielRuf** [comment]
Well, why not puppeteer and install it as peerDependency, create a postinstall script (combined with env variables) or mock it?

https://github.com/DanielRuf/phantomjs-mocked

Removing is not a solution ;-)

You give user no option to choose / decide. So it's broken and backwards incompatible which requires a new major version.

**Comment by rainfore** [comment]
Yes, I agree with you. Breaking is not good.
My loader is just a choice.

**Comment by ahmadalfy** [comment]
Bumping this up. This already seems to be resolved by some PRs.

**Comment by DanielRuf** [comment]
> Bumping this up. This already seems to be resolved by some PRs.

Can you provide more details?

**Comment by ahmadalfy** [comment]
> Can you provide more details?

This PR https://github.com/jkphl/svg-sprite/pull/284

**Comment by ddanier** [comment]
Just a note: Currently svg-sprite is not usable on Apple M1 as there is no arm64 version of phantomjs-prebuilt. I would vote for solving this soon.

**Comment by niksy** [comment]
@ddanier are you sure? Because I’m using it just fine on native M1. Linux distributions don’t have it (e.g. if you use it inside Docker), so you need to compile it from source.

**Comment by Kreeg** [comment]
@ddanier @niksy hello. It's definitely not working on M1 but on docker image of any linux dist
`#18 6.088 npm WARN deprecated phantomjs-prebuilt@2.1.16: this package is now deprecated
...
#18 10.50 npm ERR! code 1
#18 10.50 npm ERR! path /project/server/shop/static/builder/node_modules/phantomjs-prebuilt
#18 10.51 npm ERR! command failed
#18 10.51 npm ERR! command sh -c node install.js
#18 10.51 npm ERR! PhantomJS not found on PATH
#18 10.51 npm ERR! Unexpected platform or architecture: linux/arm64
#18 10.51 npm ERR! It seems there is no binary available for your platform/architecture` 

I am sure it's kind of critical due to start of arm64 arch raising. 
Adding the fact of phantomjs's 4-year deprecation status makes it even more critical

**Comment by ddanier** [comment]
@niksy Pretty sure - although it seems to work when you have rosetta. See below:

I initially tried to use this in a docker container with arch set to linux/arm64/v8. The npm/yarn install fails with the error message mentioned by @Kreeg. So here this just does not work like I originally mentioned.

Somehow the npm/yarn install done directly on the mac works when rosetta is installed. Not sure about the magic happening in the background but it is then started as an Intel process (see Activity Monitor for reference). I think npm/yarn will just allow intel packages to be installed if you have rosetta.  
(Note: I did not try to remove rosetta to verify this, so the part about this only working in this case is just my best guess. I remember trying to install it and it failed....so I expect this to be related to me now having rosetta cause of some other desktop software which required it. Feel free to really test this. I need to use svg-sprite inside docker, so this doesn't really help in my case.)

This might mean svg-sprite will work for many cases, but is for sure not a good solution overall...and just shows the technical debt you have using phantomjs pretty clearly. If you try to use it within a docker container it just does not work - people like me are just left in the woods with the current situation.

So in summary: It works on Apple M1 directly if you have rosetta. It does not work inside arm64 docker containers which will break many setups for sure.

**Comment by DanielRuf** [comment]
@ddanier basically it depends on the buildtools on your Mac and which platform they support. Normally node-gyp builds from source using Python and the commandline and developer tools (XCode and CLT) of MacOS which requires the correct libs and headers from the MacOS SDKs.

https://github.com/nodejs/node-gyp/blob/master/macOS_Catalina.md
https://mycyberuniverse.com/en-gb/how-fix-node-gyp-rebuild-fail-no-xcode-clt.html

If the relevant headers and tools are missing, the build may fail. Rosetta itself is the old / legacy API wrapper to suppport old architectures (the emulator layer).

**Comment by niksy** [comment]
@ddanier yeah, it’s probably Rosetta, haven’t thought about that. As for Docker container, we’ve built from source (CentOS 7) so we fixed it for our company project, but I agree it’s not ideal since ARM will definitely be used more over time.

**Comment by Kreeg** [comment]
Any updates?

**Comment by millette** [comment]
Is the goal only to handle svg conversion? If that's the case, perhaps <https://github.com/yisibl/resvg-js> is a better option.
